### PR TITLE
Variable: improve our `hasEquivalentVariable()` method (#321)

### PR DESCRIPTION
**Pull request by agarny**
Addresses issue #321.

### CODE REVIEW

File: src/api/libcellml/variable.h
Comment (hsorby) [inline src/api/libcellml/variable.h]: This is exactly the same as `hasEquivalentVariable`?? We don't need both just a properly working one.

Comment (agarny) [reply src/api/libcellml/variable.h]: Oops, my bad, I indeed completely overlooked your ["old" comment](https://github.com/cellml/libcellml/issues/321#issuecomment-491995223).

Ok, I am going to remove the contents of the "old" `hasEquivalentVariable()` and replace it with my new one.

Comment (hsorby) [reply src/api/libcellml/variable.h]: Sounds good.

File: src/api/libcellml/variable.h
Comment (hsorby) [inline src/api/libcellml/variable.h]: I don't think this belongs in the API documentation.  I would prefer it if we marked the method as deprecated instead.

File: src/component.cpp
Comment (hsorby) [inline src/component.cpp:37]: resets before variables?

**Review by hsorby — COMMENTED** [review]
Not keen on removing the documentation from the private methods I think this is worthwhile keeping.

File: src/api/libcellml/variable.h
Comment (agarny) [reply src/api/libcellml/variable.h]: Good point, I hadn't thought of that special command. Consider it done.

File: src/component.cpp
Comment (agarny) [reply src/component.cpp:37]: Indeed. Consider it corrected.

**Comment by agarny** [comment]
> Not keen on removing the documentation from the private methods I think this is worthwhile keeping.

It's just that we are not being consistent throughout and, I must confess, that I find it to get in the way. So, I thought I would "clean up" all the private implementation declarations by removing their corresponding documentation...?

**Comment by nickerso** [comment]
Definitely wouldn't pollute a PR with all the unrelated changes by removing the implementation documentation. Best to keep the PR focused on the issue. Cleaning up (or not) documentation of the implementation classes should be a separate issue.

**Comment by agarny** [comment]
> Definitely wouldn't pollute a PR with all the unrelated changes by removing the implementation documentation. Best to keep the PR focused on the issue. Cleaning up (or not) documentation of the implementation classes should be a separate issue.

Fair enough. I am going to put the private documentation back in...

**Comment by agarny** [comment]
Ok, I believe this is now ready.

File: src/api/libcellml/variable.h
Comment (nickerso) [inline src/api/libcellml/variable.h:219]: do we need to document what we mean by "equivalent" or add a see-also link to somewhere that is already defined?

**Review by nickerso — DISMISSED** [review]
minor comment on the documentation (which can be addressed separately), but otherwise this looks good and the tests are demonstrating the behaviour that I would expect.

File: src/api/libcellml/variable.h
Comment (nickerso) [reply src/api/libcellml/variable.h:219]: happy for this to be bumped to a separate issue.

Comment (agarny) [reply src/api/libcellml/variable.h:219]: I would rather have this addressed in a different issue.

**Comment by agarny** [comment]
FWIW, PR #343 should be merged in before this one (I still need to "fix" a couple of things in PR #343 first though).

**Comment by agarny** [comment]
@hsorby, am I correct in assuming that, for some reasons, the coverage test didn't work properly and that I probably ought to retrigger the commit?

**Comment by hsorby** [comment]
Something has gone wrong??? I'll have a look.

**Comment by agarny** [comment]
> Something has gone wrong??? I'll have a look.

Well, yes:

```
------------------------------------------------------------------------------
                           libCellML Code Coverage Report
Directory: /home/libcellml/bbslave/Coverage_Tests_Builder_Ubuntu_18_04_64_Bit/libcellml-build/src/CMakeFiles/cellml.dir
------------------------------------------------------------------------------
File                                       Lines    Exec  Cover   Missing
------------------------------------------------------------------------------
------------------------------------------------------------------------------
TOTAL                                          0       0    --%
------------------------------------------------------------------------------
```

I redelivered the commit earlier, but no difference.

**Comment by agarny** [comment]
@nickerso and @hsorby: coverage is working once again (thanks @hsorby!), so you are (very) kindly invited to review and merge this PR in... :)

**Comment by agarny** [comment]
Ping @hsorby.

File: tests/variable/variable.cpp
Comment (hsorby) [inline tests/variable/variable.cpp:395]: This test is way to verbose.  It should just do the minimal required to check for the direct equivalence.  This doesn't require 5 variables.

File: tests/variable/variable.cpp
Comment (hsorby) [inline tests/variable/variable.cpp:424]: Same as direct equivalent test too verbose.

File: tests/variable/variable.cpp
Comment (agarny) [reply tests/variable/variable.cpp:395]: Agreed, 3 variables are good enough. Addressed in 941c794.

File: tests/variable/variable.cpp
Comment (agarny) [reply tests/variable/variable.cpp:424]: Agreed, 3 variables are good enough. Addressed in 941c794.

**Comment by agarny** [comment]
@hsorby, @nickerso and @kerimoyle... feel free to (re)review this PR... :) (@kerimoyle, I can't put you as a reviewer for now, but feel free to add yourself once you have agreed to be part of the "team".)

**Comment by kerimoyle** [comment]
Am I right in thinking that when this goes through the method of variable validation should change? And how does the stuff which @hsorby is working on in #350 fit in with this? Sorry to come late to the conversation, just trying to figure out how all the bits work together ...

**Comment by agarny** [comment]
Yes, basically whenever a PR gets merged in to `develop`, you *must* merge it in your own PRs and make sure that everything is as you think it should be. It's not necessarily only a matter of merging in and resolving conflicts (should there be any). It may also involve making sure that the new code looks and behaves the way you think it should.

So, when it comes to PRs, the first one that gets merged in "wins" and the others have to be updated so that they can be safely merged afterwards.

As it happens, @nickerso modified the repository's settings to enforce that rule.

**Review by nickerso — DISMISSED** [review]
In general it makes it hard to review the actual new code when there are a bunch of unrelated code cleaning commits mixed into a PR like this. Would be good to keep them separated (and help with more speedy reviews).

**Comment by agarny** [comment]
Agreed, I did that sometime ago and, yesterday, I was wondering whether I should have removed it...

**Comment by kerimoyle** [comment]
Sorry @agarny for the delay - my questions wasn't about the mechanics of merging, it was about the intention of this branch and what @hsorby is trying to do on the other variable-related branch at PR #350 ...?

**Comment by agarny** [comment]
Ok, I shall then leave it to him to get back to you although it might probably take some time since we have a big deadline coming up this Friday...

**Comment by agarny** [comment]
Ping @hsorby, @nickerso & @kerimoyle.

**Comment by kerimoyle** [comment]
Hi again @agarny :) I'm still not clear on the changes that @hsorby is making to variables, which is what this PR and the other variable-related ones (#350) are waiting for ...

**Review by nickerso — APPROVED** [review]

**Comment by nickerso** [comment]
> Hi again @agarny :) I'm still not clear on the changes that @hsorby is making to variables, which is what this PR and the other variable-related ones (#350) are waiting for ...

This PR is focused (mostly) on addressing the need to distinguish between when a variable is directly connected to another variable versus the case when two variables are in the same set of connected variables (as we do not represent connections in the libCellML object model). Which is different to the issue @hsorby has mentioned on #350, which is a more fundamental issue with the `Variable` class definition itself. I think, its not completely clear to me :)

**Comment by nickerso** [comment]
and maybe worth pointing out that we mark this direct equivalence method as deprecated from the start as we know this is not the ideal way to achieve this, but it is the pragmatic solution to allow libCellML to progress.

**Comment by kerimoyle** [comment]
... wait ... now I'm more confused ... so there's a difference in libcellml interpretation between these two situations?

```
a is equivalent to b
b is equivalent to c
```

and

```
a is equivalent to b
a is equivalent to c
```

If so I need to rethink some stuff ... I've been treating them as all being part of the same set, no matter how/which direction the equivalence was specified from?

**Comment by nickerso** [comment]
No, you are treating them correctly.

The initial issue that brought this up was that previously libCellML would have incorrectly said that `a` is not equivalent to `c` - fixing that raised an edge case in printing and validating models where you actually need to know about the `connection` elements in the CellML XML - which are not captured in the libCellML data model. So the idea of a "direct equivalence" is a hack to know that there is an actual connection between two variables.

In general, it should be sufficient to know that variables are in the same equivalence set regardless of how many layers of connections may exist between them.

**Comment by kerimoyle** [comment]
So the methods below are there to maintain the same functionality as was previously there (albeit under different function names), but are deprecated and not really expected to be used?  ... in which case, why not just remove them completely?

Or will there ever be a situation where the direct connection is important? ... in which case, what is it?

```
bool Variable::hasDirectEquivalentVariable(const VariablePtr &equivalentVariable) const
{
    return mPimpl->hasDirectEquivalentVariable(equivalentVariable);
}

bool Variable::VariableImpl::hasDirectEquivalentVariable(const VariablePtr &equivalentVariable) const
{
    auto it = findEquivalentVariable(equivalentVariable);
    if (it == mEquivalentVariables.end()) {
        return false;
    }
    return !it->expired();
}
```

**Comment by nickerso** [comment]
Sorry, slowly coming back to me :-) the need for the direct connection
method will always exist but it's presence in the public API is deprecated
and shouldn't be used outside libcellml. The current object model forces us
to include this in the public API for now in order to use this method
internally.

The most obvious need is when printing a model to CellML XML. The printer
needs to know how and when to serialise connection elements.

**Comment by kerimoyle** [comment]
OK, got it.  Thanks!

**Review by kerimoyle — APPROVED** [review]
Looks ok to me.  (At the lowest possible level of importance, it would be good to decide what structure our comments have, if any.  For eg, there are instances of "Test ... " and "Tests ..." here in the same function definition.  A job for another PR after all the upcoming stuff is finally merged in!)

**Comment by agarny** [comment]
Two people approved, so one of you can feel free to merge this in... :)

**Pull request merged**
